### PR TITLE
Use ragged tensors for training model

### PR DESCRIPTION
**Pull request by jasonrute**
This PR is one initial step into speeding up and improving memory efficiency in the model.  The major changes are as follows:
* The training model now outputs a tensor of shape `[batch_size, None(num_args), None(context)]`.  In other words the arguments and context dimension are now ragged where the arguments dimension only ranges over argument positions used for that tactic, and the context dimension only ranged over elements in the local context or dynamic global context.
* The ground truth global arguments given to the model are indexed into the dynamic global context now.  This matches the way they are outputted by the training model.
* The losses and metrics are changed accordingly.  There are still some inefficiencies (especially in the accuracy calculations) where I convert a ragged tensor of logits back into a non-ragged tensor.  That can be fixed a bit later, but I want to start working on the main bottleneck, which is the linear operation.

This PR is a first step to speeding up the model.  Now that we have a foothold with the ragged tensor, future PRs can address the following:
- [ ] Making the linear operation (matrix multiplication) which creates the logits more efficient (but only performing the calculation over the dynamic global context and the valid argument positions)
- [ ] Consolidating the code for the local and global arguments since they are now using the same idea
- [ ] Remove any dependencies on `_valid_indices` which is no longer needed.
- [ ] Clean up the accuracy calculations to not convert back to non-ragged tensors first.
- [ ] Clean up the prediction calculation to not convert back to non-ragged tensors first.

### CODE REVIEW

**Comment by jasonrute** [comment]
The tests all pass (and were very helpful in finding some subtle bugs).

**Comment by jasonrute** [comment]
Yes.  The output of the prediction network hasn't been changed in this PR.  I'm slowing modifying everything systematically.
